core/tests/puppeteer-acceptance-tests/functions/is-element-clickable.ts

```diff
@@ -0,0 +1,157 @@
+// Copyright 2024 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Utility function to check if an element is clickable.
+ */
+
+export default function isElementClickable(element: Element): boolean {
+  /**
+   * This function gets the element that is overlapping the given element
+   * by getting the element at the center of the given element.
+   */
+  const getOverlappingElement = (
+    element: Element,
+    parent: ShadowRoot | Document = document
+  ): Element | null => {
+    const elementDimensions = element.getBoundingClientRect();
+    const x = elementDimensions.left + element.clientWidth / 2;
+    const y = elementDimensions.top + element.clientHeight / 2;
+
+    return parent.elementFromPoint(x, y);
+  };
+
+  /**
+   * This function also gets the element that is overlapping the given element
+   * but using the center of the first client rectangle of the element. Which
+   * is applicable in special cases where the text is multiline.
+   */
+  const getOverlappingRect = (
+    element: Element,
+    parent: ShadowRoot | Document = document
+  ): Element | null => {
+    const rects = element.getClientRects();
+    const rect = rects[0];
+    const x = rect.left + rect.width / 2;
+    const y = rect.top + rect.height / 2;
+
+    return parent.elementFromPoint(x, y);
+  };
+
+  /**
+   * This function combines the overlapping element and overlapping rect
+   * to get all the overlapping elements.
+   */
+  const getOverlappingElements = (
+    element: Element,
+    parent: ShadowRoot | Document = document
+  ): Element[] => {
+    const overlappingElement = getOverlappingElement(element, parent);
+    const overlappingRect = getOverlappingRect(element, parent);
+    const overlappingElements: Element[] = [];
+    if (overlappingElement) {
+      overlappingElements.push(overlappingElement);
+    }
+    if (overlappingRect) {
+      overlappingElements.push(overlappingRect);
+    }
+    return overlappingElements;
```

Comment:
This is just getting the top most overlapping element since you can't return multiple elements from a single point. We are simply getting the element dimensions from either its bounding client rect or the first client rect which is applicable for places where there are multilines like spans. Later on we recursively check for the overlapping element so if one of the top elements vanishes, then this one returns the newly "top" most element.

Reply:
OK, thanks. In that case, should the docstring at the top say "topmost" somewhere?

Also, I'm not sure what you mean by "only checks if the element is currently clickable not in the future or more overlapping elements,", but I think I get the general gist. Maybe you can explain more in the docstring, e.g. that it only fetches the topmost element of each type, and the method name can be something like getTopmostOverlappingElements perhaps.